crates/dbsp/src/operator/dynamic/input_upsert.rs

```diff
@@ -610,26 +689,62 @@ where
                 cur_key.from_ref(key);
                 cur_val.set_none();
 
-                // Generate retraction if `key` is present in the trace.
-                if trace_cursor.seek_key_exact(key, None) {
-                    // println!("{}: found key in trace_cursor", Runtime::worker_index());
-                    while trace_cursor.val_valid() {
-                        let weight = **trace_cursor.weight();
+                // If `key` is in the in-memory trace, then get it from the
+                // overall trace and generate a retraction.
+                //
+                // Otherwise, it might be in the on-disk trace, so append it to
+                // `keys_builder`.
+                enum Action {
+                    GenerateRetraction,
+                    FetchFromStorage,
+                }
+                let action = match (memory_batches.is_empty(), storage_batches.is_empty()) {
```

Comment:
these branches are always taken the same way, so this loop could be rewritten as 4 different loops by checking this first. not sure if there would be a performance benefit. Most likely not when storage is involved.